DPPP/MSBDAReader.h

```diff
@@ -0,0 +1,203 @@
+// Copyright (C) 2020
+// ASTRON (Netherlands Institute for Radio Astronomy)
+//
+// This file is part of the LOFAR software suite.
+// The LOFAR software suite is free software: you can redistribute it and/or
+// modify it under the terms of the GNU General Public License as published
+// by the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The LOFAR software suite is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
+
+/// @file
+/// @brief Step for reading BDA data from an MS.
+/// @author Lars Krombeen
+
+#ifndef DPPP_MSBDAREADER_H
+#define DPPP_MSBDAREADER_H
+
+#include "DPInput.h"
+#include "BDABuffer.h"
+#include "UVWCalculator.h"
+#include "FlagCounter.h"
+
+#include <casacore/tables/Tables/TableIter.h>
+#include <casacore/tables/Tables/RefRows.h>
+#include <casacore/casa/Arrays/Slicer.h>
+
+#include <map>
+
+namespace DP3 {
+
+class ParameterSet;
+
+namespace DPPP {
+/// @brief Step for reading BDA data from an MS.
+
+/// This class is a DPInput step reading the data from a MeasurementSet.
+/// At the beginning it finds out the shape of the data; i.e., the
+/// number of correlations, channels, baselines, and time slots.
+/// It requires the data to be regularly shaped.
+///
+/// The object is constructed from the 'msin' keywords in the parset file.
+/// Currently the following can be given:
+/// <ul>
+///  <li> msin: name of the MS
+///  <li> msin.datacolumn: the data column to use [DATA]
+///  <li> msin.weightcolumn: the weights column to use [WEIGHT_SPECTRUM or
+///           WEIGHT]
+/// </ul>
+///
+/// The process function only reads the data and flags to avoid that
+/// too much data is kept in memory.
+/// Other columns (like WEIGHT, UVW) can be read when needed by using the
+/// appropriate DPInput::fetch function.
```

Comment:
This comment is incorrect: The BDA reader always reads weights and UVW. Moreover, the fetch functions only work with regular buffers, not with BDA buffers.